src/mozanalysis/experiment.py

```diff
@@ -534,3 +464,161 @@ def _get_telemetry_sanity_check_metrics(self, enrollments, data_source):
                 & F.isnull(data_source.experiments[self.experiment_slug])
             ).astype('int'), F.lit(0))).alias('has_non_enrolled_data'),
         ]
+
+
+@attr.s(frozen=True)
+class TimeLimits(object):
+    """Internal object containing various time limits.
+
+    Instantiated and used by the ``Experiment`` class; end users
+    should not need to interact with it.
+
+    Do not directly instantiate: use ``TimeLimits.create()``.
+
+    There are several time constraints needed to specify a valid query
+    for experiment data:
+
+        * When did enrollments start?
+        * When did enrollments stop?
+        * How long after enrollment does the analysis window start?
+        * How long is the analysis window?
+
+    Even if these four quantities are specified directly, it is
+    important to check that they are consistent with the available
+    data - i.e. that we have data for the entire analysis window for
+    every enrollment.
+
+    Furthermore, there are some extra quantities that are useful for
+    writing efficient queries:
+
+        * What is the first date for which we need data from our data
+          source?
+        * What is the last date for which we need data from our data
+          source?
+
+    Instances of this class store all these quantities and do validation
+    to make sure that they're consistent. The "store lots of overlapping
+    state and validate" strategy was chosen over "store minimal state
+    and compute on the fly" because different state is supplied in
+    different contexts.
+    """
+    first_enrollment_date = attr.ib(type=str)
+    last_enrollment_date = attr.ib(type=str)
+
+    analysis_window_start = attr.ib()
+    """The integer number of days between enrollment and the start of
+    the analysis window, represented either as an ``int`` or as a
+    Column of integers in the ``enrollments`` DataFrame."""
```

Comment:
I avoided `F.lit(0)` because I suspected that Spark might treat a literal literal differently to a column literal, and because I wanted to do maths in Python with it where possible (see `_validate_analysis_window_start()`)